### PR TITLE
JOSS paper about Pooch 

### LINKED ISSUE #112: JOSS paper
I hadn't been considering getting a JOSS paper for Pooch because I thought it wouldn't qualify as "research software". But I just noticed that [tqdm has a JOSS paper](https://joss.theoj.org/papers/10.21105/joss.01277), so I think we can make a case for Pooch. After all, other research software would cite Pooch in their papers (ideally). **What do others think? Is this worth pursuing?**

The criteria for authorship that seems reasonable to me would be:

1. Have contributed code or documentation to Pooch beyond typo fixes 
2. Read/edit the JOSS paper and give their :+1:

This would basically mean anyone [listed on Github](https://github.com/fatiando/pooch/graphs/contributors) that at least reads and OKs the paper. 

As for the order, we could go with the GitHub contributor order (I think it's by number of commit?), and do alphabetical for a tie break. But I'm not entirely sure how to proceed with this one. Any input would be welcome.

Right now, potential authors would be me @santisoler @remram44 @hugovk @matthewturk @jrleeman @matthewturk @danshapero. If any of you are interested, please reply to this thread with your consent, affiliation, and ORCID. Also, whether you think it's worth submitting (the paper itself will take ~1h to write copying from our docs).

### PR DESCRIPTION
**Pull request by leouieda**
This is the paper we plan to submit to the Journal of Open Source Software. 
The submission will be for version 0.6 but I expect we can release 1.0.0 by the time the paper is accepted.

Fixes #112 

@santisoler @remram44 @hugovk @matthewturk @jrleeman @danshapero I'd appreciate feedback on the paper, in particular to make sure we're clearly meeting all of the requirements from JOSS (see [their docs for details](https://joss.readthedocs.io/en/latest/submitting.html)):

1. The software should have an **obvious** research application.
3. The software should be a significant contribution to the available open source software that either enables some new research challenges to be addressed or makes addressing research challenges significantly better (e.g., faster, easier, simpler).
4. The software should be feature-complete (no half-baked solutions) and designed for maintainable extension (not one-off modifications). **Minor ‘utility’ packages, including ‘thin’ API clients, are not acceptable**.

I tried to make it clear in the text the *research application*, how we make research easier, and that we're not a minor utility. I'm sure it could be better, though. These are the only reasons the paper would be considered for rejection so we have to make sure we're clearly meeting these requirements.

Also, please feel free to add any references or packages using Pooch that I missed.

### CODE REVIEW

**Comment by leouieda** [comment]
This is a preview of the paper PDF: http://res.cloudinary.com/hju22ue2k/image/upload/v1572880531/hubnwkesoorklcrjx7fu.pdf

You can recompile it using https://whedon.theoj.org

**Comment by leouieda** [comment]
@danshapero I included a citation to icepack. The Zenodo DOI archive has the title "icepack/icepack: Alpha release for citation purposes", which I suspect is the title of the GitHub release. Would this be something you want to change?

**Comment by danshapero** [comment]
@leouieda Hah yes thanks for the suggestion :)

File: paper/paper.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+   index: 6
+ - name: Polar Science Center, University of Washington Applied Physics Lab, USA
+   index: 7
+date: 23 October 2019
+bibliography: paper.bib
+---
+
+# Summary
+
+Scientific software are usually created to analyze, model, and visualize data.
```
Comment (remram44) [inline paper/paper.md]: software is uncountable
```suggestion
Scientific software is usually created to analyze, model, and visualize data.
```
or maybe "scientific software packages/products/projects"

File: paper/paper.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+As such, many software libraries include sample datasets in their distributions
+for use in documentation, tests, benchmarks, and workshops.
+Prominent examples in Python include scikit-learn [@scikit-learn] and
+scikit-image [@scikit-image].
+The usual approach is to include smaller datasets in the GitHub repository
+directly and package them with the source and binary distributions.
+Larger datasets require writing code to download the files from a remote server
+to the users computer.
```
Comment (remram44) [inline paper/paper.md]:
```suggestion
to the user's computer.
```

File: paper/paper.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+Pooch is designed to be extended: users can plug-in custom download functions
+and post-download processing functions.
+For example, a custom download function could fetch files over FTP instead of
+HTTP (the default) and a processing function could decrypt a file using a
+user-defined password once the download is completed.
+We include ready-made download functions for HTTP (including basic
+authentication) and processing functions for unpacking archives (zip or tar)
+and decompressing files (gzip, lzma, and bzip2).
+
+To the best of the authors awareness, the only other Python software with some
```
Comment (remram44) [inline paper/paper.md]:
```suggestion
To the best of the authors' awareness, the only other Python software with some
```

File: paper/paper.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+It manages a data *registry* by downloading files from one or more remote
+servers and storing them in a local data cache.
+Pooch is written in pure Python and has minimal dependencies.
+The integrity of downloads is verified by comparing the file's SHA256 hash with
+the one stored in the data registry.
+This is also the mechanism used to detect if a file needs to be re-downloaded
+due to an update in the registry.
+Pooch is meant to be a drop-in replacement for the custom download code that
+users have already written (or are planning to write).
+In the ideal scenario, the end user of a software should not need to know that
```
Comment (remram44) [inline paper/paper.md]:
```suggestion
In the ideal scenario, the end-user of a software package should not need to know that
```

File: paper/paper.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+To the best of the authors awareness, the only other Python software with some
+overlapping functionality is [Intake](https://github.com/intake/intake).
+While Intake is powerful and can be used to manage large data archives,
+we argue that Pooch is has a simpler setup and meets the
+specific needs of scientific software authors and individual scientists.
+For example, Pooch does not require users to change their data loading code to
+fit into its plug-in structure, instead only providing the file path for the
+user.
+
+The Pooch API is stable and has been field tested by other projects:
```
Comment (remram44) [inline paper/paper.md]:
```suggestion
The Pooch API is stable and has been field-tested by other projects:
```

**Comment by leouieda** [comment]
> Looks good! Some minor fixes

Thanks, @remram44! Those pesky `'` and `-` get me every time.

File: README.rst [hunk 6]
```diff
@@ -31,9 +31,40 @@ Part of the `Fatiando a Terra <https://www.fatiando.org>`__ project
...
+Here are Pooch's main features:
+
+* Pure Python and minimal dependencies.
+* Download a file only if necessary (it's not in the data cache or needs to be updated).
+* Verify download integrity through SHA256 hashes (also used to check if a file needs to
+  be updated).
+* Designed to be extended: plug-in custom download (FTP, scp, etc) and post-processing
+  (unzip, decompress, rename) functions.
+* Includes utilities to unzip/decompress the data upon download to save loading time.
+* Can handle basic HTTP authentication (for servers that require a log in) and printing
```
Comment (hugovk) [inline README.rst]: "login" is the noun form:
```suggestion
* Can handle basic HTTP authentication (for servers that require a login) and printing
```

File: README.rst [hunk 7]
```diff
@@ -31,9 +31,40 @@ Part of the `Fatiando a Terra <https://www.fatiando.org>`__ project
...
+  download progress bars.
+* Easily setup an environment variable to overwrite the data cache location.
```
Comment (hugovk) [inline README.rst]: "set up" is the verb form:
```suggestion
* Easily set up an environment variable to overwrite the data cache location.
```

File: paper/paper.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+for use in documentation, tests, benchmarks, and workshops.
+Prominent examples in Python include scikit-learn [@scikit-learn] and
+scikit-image [@scikit-image].
+The usual approach is to include smaller datasets in the GitHub repository
+directly and package them with the source and binary distributions.
+Larger datasets require writing code to download the files from a remote server
+to the user's computer.
+The same problem is faced by scientists using version control to manage their
+research projects.
+As data files increase in size, it becomes unfeasible to store them on GitHub
```
Comment (hugovk) [inline paper/paper.md]:
```suggestion
As data files increase in size, it becomes unfeasible to store them in GitHub
```

File: paper/paper.md [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+repositories.
+While downloading a data file over HTTP can be done easily with modern Python
+libraries, it is not trivial to manage a set of files, keep them updated, and
+check for corruption.
+Instead of scientists and library authors recreating the same code, it would be
+best to have a minimalistic and easy to setup tool for fetching and maintaining
```
Comment (hugovk) [inline paper/paper.md]:
```suggestion
best to have a minimalistic and easy to set up tool for fetching and maintaining
```

File: paper/paper.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+due to an update in the registry.
+Pooch is meant to be a drop-in replacement for the custom download code that
+users have already written (or are planning to write).
+In the ideal scenario, the end-user of a software package should not need to know that
+Pooch is being used.
+Setup is as easy as calling a single function (`pooch.create`), including
+setting up an environment variable for overwriting the data cache path and
+versioning the downloads so that multiple versions of the same package can
+coexist in the same machine.
+For example, this is the code required to setup a module
```
Comment (hugovk) [inline paper/paper.md]:
```suggestion
For example, this is the code required to set up a module
```

File: paper/paper.md [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+# Get the version string from the project
+from . import version
+
+# Create a new instance of pooch.Pooch
+GOODBOY = pooch.create(
+    # Cache path using the default for the operating system
+    path=pooch.os_cache("myproject"),
+    # Base URL of the remote data server (for example, on GitHub)
+    base_url="https://github.com/me/myproject/raw/{version}/data/",
+    # PEP440 compliant version number (added to path and base_url)
```
Comment (hugovk) [inline paper/paper.md]:
```suggestion
    # PEP 440 compliant version number (added to path and base_url)
```

File: paper/paper.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+)
+# Load the registry from a simple text file.
+# Each line has: file_name sha256 [url]
+GOODBOY.load_registry("registry.txt")
+
+def fetch_some_data():
+    # Get the path to the data file in the local cache
+    # If it's not there or needs updating, download it
+    fname = GOODBOY.fetch("some-data.csv")
+    # Load it with numpy/pandas/xarray/etc
```
Comment (hugovk) [inline paper/paper.md]:
```suggestion
    # Load it with NumPy/pandas/xarray/etc.
```

File: README.rst [hunk 13]
```diff
@@ -31,9 +31,40 @@ Part of the `Fatiando a Terra <https://www.fatiando.org>`__ project
...
+from a server only when needed and storing them locally in a data *cache* (a folder on
+your computer).
+
+Here are Pooch's main features:
+
+* Pure Python and minimal dependencies.
+* Download a file only if necessary (it's not in the data cache or needs to be updated).
+* Verify download integrity through SHA256 hashes (also used to check if a file needs to
+  be updated).
+* Designed to be extended: plug-in custom download (FTP, scp, etc) and post-processing
```
Comment (hugovk) [inline README.rst]:
```suggestion
* Designed to be extended: plug in custom download (FTP, scp, etc) and post-processing
```

File: paper/paper.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    data = pandas.read_csv(fname)
+    return data
+```
+
+Pooch is designed to be extended: users can plug-in custom download functions
```
Comment (hugovk) [inline paper/paper.md]:
```suggestion
Pooch is designed to be extended: users can plug in custom download functions
```

File: paper/paper.md [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+HTTP (the default) and a processing function could decrypt a file using a
+user-defined password once the download is completed.
+We include ready-made download functions for HTTP (including basic
+authentication) and processing functions for unpacking archives (zip or tar)
+and decompressing files (gzip, lzma, and bzip2).
+
+To the best of the authors' awareness, the only other Python software with some
+overlapping functionality is [Intake](https://github.com/intake/intake).
+While Intake is powerful and can be used to manage large data archives,
+we argue that Pooch is has a simpler setup and meets the
```
Comment (hugovk) [inline paper/paper.md]:
```suggestion
we argue that Pooch has a simpler setup and meets the
```

**Comment by hugovk** [comment]
> Also, please feel free to add any references or packages using Pooch that I missed.

Clicking the number next to "Used by" at the top right of the repo shows a dependency graph.

Here's the dependent [packages](https://github.com/fatiando/pooch/network/dependents?dependent_type=PACKAGE&package_id=UGFja2FnZS0xMDIyMTgwMjQ%3D) (those in bold already mentioned):

* **fatiando/rockhound**
* fatiando/harmonica
* **aashish24/MetPy**
* **fatiando/verde**

And [repositories](https://github.com/fatiando/pooch/network/dependents?dependent_type=REPOSITORY&package_id=UGFja2FnZS0xMDIyMTgwMjQ%3D) (duplicates removed):

* apottr/nexrad-something
* LandscapeGeoinformatics/EstSoil-EH_sw_supplement
* apottr/nexrad-process
* frontogenesis/flask
* metwork-framework/mfextaddon_scientific
* kuchaale/X-regression
* **icepack/icepack**
* JoDeMiro/Eniko
* leouieda/test-doctr
* JustinGOSSES/predictatops
* cropsinsilico/hothouse
* meteorolog90/PROGRAM
* **Unidata/MetPy**

I doubt they all need mentioning, but worth a look?

File: README.rst (hunk 7 above)
Comment (leouieda) [reply README.rst]: Thanks! I didn't know that.

File: paper/paper.md (hunk 12 above)
Comment (leouieda) [reply paper/paper.md]: Leaving this one out since all other comments don't have periods. Should I include one in all sentences?

Comment (hugovk) [reply paper/paper.md]: Sure, I only added it for the "etc" abbreviation, but don't worry about it for comments :)

Comment (leouieda) [reply paper/paper.md]: Ah, right sorry I missed that!

**Comment by leouieda** [comment]
@hugovk the GitHub dependent graph is kind of strange and not very accurate. I was looking at these now and most are false positives. Some get Pooch through a dependency on MetPy. The main ones I missed were [predictatops](https://github.com/JustinGOSSES/predictatops) and [hothouse](https://github.com/cropsinsilico/hothouse) (which seems to not have been release yet but @matthewturk could tell us more).

[aletheia-data](https://github.com/NCAR/aletheia-data) is actually a fork of Pooch adding ftp downloads so I'm sure this counts as "using Pooch". I've opened NCAR/aletheia-data#4 asking the author to contribute the code back to Pooch so others can benefit.

File: paper/paper.md [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+---
+
+# Summary
+
+Scientific software is usually created to analyze, model, and visualize data.
+As such, many software libraries include sample datasets in their distributions
+for use in documentation, tests, benchmarks, and workshops.
+The usual approach is to include smaller datasets in the GitHub repository
+directly and package them with the source and binary distributions
+(e.g., scikit-learn [@scikit-learn] and scikit-image [@scikit-image] do this).
```
Comment (leouieda) [inline paper/paper.md]: Rephrased this part but I'm sure I like it. Thoughts?

**Comment by leouieda** [comment]
Latest PDF rendering here: http://res.cloudinary.com/hju22ue2k/image/upload/v1572943332/tvzkzrgmorkw6lcycx46.pdf

**Comment by leouieda** [comment]
Hi @matthewturk @jrleeman @danshapero please let me know when you have some time to take a look at the paper and give a :+1: or :-1:

**Comment by matthewturk** [comment]
Sorry for not chiming in sooner -- this looks awesome, and I'm delighted to :+1: it!

**Comment by danshapero** [comment]
This looks great! Do you think it's worth waiting until the PR to to use pooch in scikit-image is merged? It might look better to be able to say that a big and popular package is already using pooch than to say that they plan to include it.

**Comment by hugovk** [comment]
Handy link to the PR to use pooch in scikit-image:

* https://github.com/scikit-image/scikit-image/pull/3945

**Comment by leouieda** [comment]
> Do you think it's worth waiting until the PR to to use pooch in scikit-image is merged? 

It would look better but I don't think that PR is getting merged too soon. It touches a lot of files and there is a lot of discussion still about some details.

**Comment by andersy005** [comment]
This is looking great! 👍

**Comment by leouieda** [comment]
Thank you for the :+1: @andersy005!

**Comment by leouieda** [comment]
Here is the latest PDF with @andersy005 added as an author and some changes to add the new FTP download capabilities: http://res.cloudinary.com/hju22ue2k/image/upload/v1574183200/npqrquwcxkewpy6ap2i5.pdf

**Comment by leouieda** [comment]
Hi everyone, I'd like to submit this paper tomorrow (2019/11/22) since it seems to be done and :+1: by everyone who agreed to be an author. Any last minute changes can still be made until then (or after during review).

**Comment by leouieda** [comment]
Submitted! :tada:

**Comment by hugovk** [comment]
And here's the submission to follow along its progress:

* https://github.com/openjournals/joss-reviews/issues/1928

**Comment by leouieda** [comment]
Thanks for the link @hugovk! We need to suggest some reviewers. Any thoughts? Maybe one of the intake developers?

**Pull request merged**